Update Database, Role and User Management

Hunterness opened the pull request:
Include information about `IF [NOT] EXIST` and `OR REPLACE` for create and drop.

Updated with more information about dropping users and roles, and other missing sections.

Hunterness committed:
Added info about IF [NOT] EXISTS and OR REPLACE to the existing commands

Hunterness committed:
Added drop role and drop user sections, including info about IF EXISTS

Hunterness committed:
Updated some of Role Management

* Added mentions of CREATE ROLE AS COPY OF
* Added section about revoke role from user
* Added mentions of granting/revoking multiple roles to/from multiple
users
* Added descriptions for the different SHOW ROLES commands
* Also commented out some println so we don't print things to std out
when running the tests

Hunterness committed:
Updated some of User Management

* Added section for changing the logged in users password
* Added description of the default values to CREATE USER

Hunterness committed:
Updated some of Database Management

* Added information about SHOW DATABASE db_name
* Added information about SHOW DEFAULT DATABASE
* Added method for asserting on one database shown and not just all
* Fixed some weird sentences

craigtaverner commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -12,10 +12,14 @@ This section explains how to use Cypher to manage Neo4j role-based access contro
 *** <<administration-security-users-show, Listing users>>
 *** <<administration-security-users-create, Creating users>>
 *** <<administration-security-users-alter, Modifying users>>
+*** <<administration-security-users-alter-password, Changing logged in users password>>
```
Would it sound nicer to say _current user's_ or _own_ instead of _logged in users_?

Hunterness replied:
I took what the table of commands said but sure I can change it :)

Hunterness committed:
Added comments about the SHOW USER/ROLE PRIVILEGES

in the user/role management sections saying that it is explained in
the listing privileges section

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DatabasesTest.scala:
```diff
@@ -30,8 +30,17 @@ class DatabasesTest extends DocumentingTest with QueryStatisticsTestSupport {
+      query("SHOW DATABASES", assertDatabasesShown) {
+        resultTable()
+      }
+      p("A particular database can be seen using the `SHOW DATABASE database_name`.")
```
I would probably shorten `database_name` to `name` as the 'database' part is very much implied from context.

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DatabasesTest.scala:
```diff
@@ -30,8 +30,17 @@ class DatabasesTest extends DocumentingTest with QueryStatisticsTestSupport {
+        resultTable()
+      }
+      p("The default database can be seen using the `SHOW DEFAULT DATABASE`.")
+      query("SHOW DATABASE neo4j", assertDatabaseShown("neo4j")) {
```
The query is wrong. Should use the term `DEFAULT`.

Hunterness replied:
yes, that is correct

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DatabasesTest.scala:
```diff
@@ -44,10 +53,19 @@ class DatabasesTest extends DocumentingTest with QueryStatisticsTestSupport {
         resultTable()
       }
+      p("This command is optionally idempotent, with the default behavior to throw an exception if the database already exists. " +
+        "Appending `IF NOT EXISTS` to the command will ensure that no exception is thrown and nothing happens should the database already exist. " +
```
Rather use the scala triple quote syntax for multi-line text.

Hunterness replied:
I think I just continued with how things already looked for this... I don't care about the text being multi-lined in itself but just that you can see the text in the code...

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/DatabasesTest.scala:
```diff
@@ -84,18 +102,34 @@ class DatabasesTest extends DocumentingTest with QueryStatisticsTestSupport {
         resultTable()
       }
+      p("This command is optionally idempotent, with the default behavior to throw an exception if the database does not exists. " +
+        "Appending `IF EXISTS` to the command will ensure that no exception is thrown and nothing happens should the database not exist.")
```
Nicer to use scala triple quotes for multi-lines.

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -37,6 +42,15 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+          assertStats(r, systemUpdates = 0)
+        })) {}
+        query("CREATE OR REPLACE USER jake SET PASSWORD 'xyz'", ResultAssertions( r => {
+          assertStats(r, systemUpdates = 2)
```
It would be good to check if the visual output includes the system updates, as it does for nodesCreate in normal Cypher. I remember this was not working last week, but I hoped it was due only to a mismatch between Neo4j and docs. If it still does not work, we should report a bug for one of us to fix in docs later.

Hunterness replied:
I have not seen any visual output for the system updates no

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -53,15 +67,58 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
           resultTable()
         }
       }
+      section("Changing the logged in users password", "administration-security-users-alter-password") {
```
I prefer the terms `current user's` or `own`.

craigtaverner replied:
Although I now see that 'current user' and 'own' are used directly below, so perhaps you wanted three ways of saying this directly after each other so the user is completely clear on what it means?

Hunterness replied:
I took the title form the description in the commands table at the start of the section, have not put any thought into word choices as such

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -53,15 +67,58 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+        query("ALTER CURRENT USER SET PASSWORD FROM 'abc123' TO '123xyz'", ResultAssertions((r) => {
+          assertStats(r, systemUpdates = 1)
+        })) {
+          // Can't pull on the result since the test is run with auth disabled, which is not allowed for this command
```
Are all system commands run with auth-disabled in the docs tests? Perhaps we need to fix that?

Hunterness replied:
I think they are, and the show commands does not show any default users/roles/privileges

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -53,15 +67,58 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+      }
     }
     section("Role Management", "administration-security-roles") {
+      initQueries("CREATE USER jake SET PASSWORD 'abc123' CHANGE NOT REQUIRED")
```
Seems like my last bug-fix worked, now you can run initQueries like this! :-)

Hunterness replied:
It was very nice since the last thing I do in the user management is dropping the user :P

Hunterness replied:
Sadly it does not always work (sometimes I get weird test failures...

Ex:
Adding `initQueries("CREATE USER user1 SET PASSWORD 'abc'", "CREATE USER user2 SET PASSWORD 'abc'", "CREATE USER user3 SET PASSWORD 'abc'", "CREATE ROLE role1", "CREATE ROLE role2")` to the start of Assigning roles to users section breaks both granting and revoking role myrole to/from jake....

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -53,15 +67,58 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+          p("This is the same command as `SHOW ALL ROLES`.")
           resultTable()
         }
+        p("There are multiple versions of this command, the default being `SHOW ALL ROLES`. " +
```
Looks better with scala triple-quote multi-line style.

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -53,15 +67,58 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
         }
+        p("There are multiple versions of this command, the default being `SHOW ALL ROLES`. " +
+          "To only show roles that are assigned to users, the command is `SHOW POPULATED ROLES`. " +
+          "To see which users are assigned to roles can `WITH USERS` be appended to the commands. " +
```
The word `can` should be moved later in the sentence.

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -88,19 +179,44 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
           resultTable()
         }
+        p("It is possible to assign multiple roles to multiple users in one command.")
+        query("GRANT ROLES role1, role2 TO user1,user2,user3", ResultAssertions( r => {
```
You have whitespace in the first list, not not the second. I think adding whitespace to the second list will look better.

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -88,19 +179,44 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+        }
+        p("It is possible to revoke multiple roles from multiple users in one command.")
+        query("REVOKE ROLES role1, role2 FROM user1,user2,user3", ResultAssertions( r => {
+          assertStats(r, systemUpdates = 4)
```
Why 4 updates, and not 6? Is it because only 4 combinations exist?

Hunterness replied:
I think the 4 comes from I had 2 roles and 2 users, it should be be 6 now (unclear why it does not fail...)

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -88,19 +179,44 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+        query("REVOKE ROLES role1, role2 FROM user1,user2,user3", ResultAssertions( r => {
+          assertStats(r, systemUpdates = 4)
+        })) {
+          // Can't pull on the result since neither roles nor users exist
```
This comment seems to contradict the `systemUpdates = 4` above?

Hunterness replied:
Since the 4 should be a 6 it might be that that is not checked either

Hunterness replied:
If I don't print the resultTable() then the statistics are not asserted on, is there a way of doing the asset while not printing the table to the docs? (I have several places with empty bodies or just printing a paragraph that does not actually test the assertion)

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -88,19 +179,44 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
     val tx = db.beginTransaction(Type.explicit, AnonymousContext.read())
     try {
-      println(p.resultAsString)
+//      println(p.resultAsString)
```
Perhaps best to just delete the lines.

Hunterness replied:
Did you have a reason for them to be here? There are several in the privilege tests as well - do we want to keep them or remove them? (they do print things when running the tests/building docs)

craigtaverner reviewed (changes requested):
Some great additions. I think you more than doubled the content. A few small suggestions and one query mistake to fix.

Hunterness committed:
Updates after building docs

* Changed some weird sentences
* Added comments about which commands `OR REPLACE` replaces
* Moved SHOW ROLES part for creating roles to be after COPY so both
the created are shown
* Tried fixing so the note in changing the logged in users password
is seen

Hunterness committed:
Fixed some of the review comments

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityUserAndRoleManagementTest.scala:
```diff
@@ -89,31 +90,33 @@ class SecurityUserAndRoleManagementTest extends DocumentingTest with QueryStatis
+          "To see which users are assigned to roles `WITH USERS` can be appended to the commands. " +
           "This will give one result row for each user, so if a role is assigned to two users then it will show up twice in the result.")
-        query("SHOW POPULATED ROLES WITH USERS", assertNodesShown("Role")) {
+        query("SHOW POPULATED ROLES WITH USERS", assertRolesShown()) {
```
This seems to assert that no roles are assigned users? Would it not be nicer to demo this command after some users are assigned roles, so we can see some results in the table?

Hunterness replied:
I added some roles and users that I needed for my commands down in `Assign roles to users` but before that nothing was returned in the `SHOW ROLES` either... If we get the default user and roles then there will be one here :)

craigtaverner reviewed (approved):
I've approved it, but I think some more changes would be good, but I'd rather make them after a review of the final built version as a second round of polish.
